herd/runTest.ml

```diff
@@ -63,6 +63,22 @@ module Make
         | Some hash ->
             TestHash.check_env env name.Name.name filename hash in
         let test = T.build name parsed in
+        let dic_pred, idc_pred =
+          let open CacheType in
+            match cache_type with
+            | None ->
+                (fun _ -> false), (fun _ -> false)
+            | Some cache_type ->
+                cache_type.dic, cache_type.idc in
+        let procs =
+          List.map (fun (p,_,_) -> p) (test.Test_herd.start_points) in
+        let all_dic = List.fold_left (fun acc p -> acc && (dic_pred p)) true procs in
+        let all_idc = List.fold_left (fun acc p -> acc && (idc_pred p)) true procs in
+        let () =
+          let any_dic = List.fold_left (fun acc p -> acc || (dic_pred p)) false procs in
+          let any_idc = List.fold_left (fun acc p -> acc || (idc_pred p)) false procs in
+          if (not all_idc && any_idc) || (not all_dic && any_dic) then
+            Warn.user_error "Each PE must have the same DIC and IDC cache type features\n"; in
```

Comment:
Good points, @relokin! One nuance: my intention was for herd7 to finish execution upon encountering this "user error" rather than display a warning and continue. Do you think that is a sensible approach? (Provided that the message is rephrased accordingly.)

Reply:
@relokin I think I might have found a graceful way of doing the check. The last commit changes the logic into "using intersection" for the DIC/IDC features and showing a warning if there is a mismatch, but otherwise proceeding. Does it look more like what you've had in mind?